tachyon/math/finite_fields/binary_prime_field.h

```diff
@@ -0,0 +1,223 @@
+// Copyright 2023 Ulvetanna Inc.
+// Use of this source code is governed by a Apache-2.0 style license that
+// can be found in the LICENSE.ulvetanna file.
+
+#ifndef TACHYON_MATH_FINITE_FIELDS_BINARY_PRIME_FIELD_H_
+#define TACHYON_MATH_FINITE_FIELDS_BINARY_PRIME_FIELD_H_
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include <limits>
+#include <optional>
+#include <string>
+#include <type_traits>
+
+#include "tachyon/base/logging.h"
+#include "tachyon/base/random.h"
+#include "tachyon/base/strings/string_number_conversions.h"
+#include "tachyon/base/strings/string_util.h"
+#include "tachyon/build/build_config.h"
+#include "tachyon/math/base/gmp/gmp_util.h"
+#include "tachyon/math/finite_fields/binary_tower_operations.h"
+#include "tachyon/math/finite_fields/prime_field_base.h"
+
+namespace tachyon::math {
+
+template <typename Config>
+class PrimeFieldGpu;
+
+template <typename _Config>
+class PrimeField<_Config, std::enable_if_t<_Config::kIsBinaryPrime &&
+                                           !_Config::kIsSpecialPrime>>
+    final : public PrimeFieldBase<PrimeField<_Config>> {
+ public:
+  constexpr static size_t kBits = _Config::kModulusBits - 1;
+  constexpr static size_t kLimbNums = (kBits + 63) / 64;
+  constexpr static size_t N = kLimbNums;
+
+  using Config = _Config;
+  using Type = std::conditional_t<
+      kBits <= 8, uint8_t,
+      std::conditional_t<kBits == 16, uint16_t,
+                         std::conditional_t<kBits == 32, uint32_t, uint64_t>>>;
+  using BigIntTy = BigInt<N>;
+  using MontgomeryTy = BigInt<N>;
+  using value_type = Type;
```

Comment:
Just asking, is there a reason for doing like this instead of
```suggestion
  using value_type = std::conditional_t<
      kBits <= 8, uint8_t,
      std::conditional_t<kBits == 16, uint16_t,
                         std::conditional_t<kBits == 32, uint32_t, uint64_t>>>;
```
?

Reply:
Since `Type` is already evaluated as such! I think what you want to ask is rather why we have an additional `Type` type..? If so, I would say it's a personal choice!